man/permuteGPAT++.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "PERMUTEGPAT++" "1" "" "permuteGPAT++ (vcflib)" "permuteGPAT++ (VCF phenotype)"
.hy
.SH NAME
.PP
\f[B]permuteGPAT++\f[R]
.SH SYNOPSIS
.PP
\f[B]permuteGPAT++\f[R] -f gpat.txt -n 5 -s 1
.SH DESCRIPTION
.PP
\f[B]permuteGPAT++\f[R] is a method for adding empirical p-values to a
GPAT++ score.
.SH OPTIONS
.IP
.nf
\f[C]


     Currently **permuteGPAT++** only supports wcFst, but will be extended.    

OUTPUT: **permuteGPAT++** will append three additional columns:
        1. The number of successes                         
        2. The number of trials                            
        3. The empirical p-value                           

file:    f   -- argument: the input file     
number:  n   -- argument: the number of permutations to run for each value [1000]
success: s   -- argument: stop permutations after \[aq]s\[aq] successes [1]

Type: phenotype
\f[R]
.fi
.SH EXIT VALUES
.TP
\f[B]0\f[R]
Success
.TP
\f[B]not 0\f[R]
Failure
.SH SEE ALSO
.PP
\f[B]vcflib\f[R](1)
.SH OTHER
.SH LICENSE
.PP
Copyright 2011-2026 (C) Erik Garrison and vcflib contributors.
MIT licensed.
Copyright 2020-2026 (C) Pjotr Prins.
.SH AUTHORS
Erik Garrison and vcflib contributors.


RELEASE_NOTES.md
For contributions see
[contributors](https://github.com/vcflib/vcflib/graphs/contributors)
and
[commits](https://github.com/vcflib/vcflib/commits/master).

## ChangeLog ongoing

* Added guix-container-manifest.scm that allows building Singularity and Docker containers(!)

## ChangeLog v1.0.15 (20260319)

* Fix iterator usage in vcfkeepinfo and vcfremoveaberrantgenotypes (Jukka Matilainen)
* Fixes for vcfwave (Glenn Hickey)
* Added include files to build with vg
* Added pkgconfig support
* Fix wfa2lib builds and sent upstream
* Update zig build to 0.15, or later, and fix static build with zig for vcfcreatemulti
* Create tar ball with static binaries
* Fix building with git submodules only
* Download singularity and docker containers from https://files.genenetwork.org/software/vcflib/v1.0.15/
  md5sum:
  5768540e7bef2db1363e422bc37614c4  vcflib-1.0.15-docker.tar.gz # Docker
  680cd7675325561636e03274e679d195  vcflib-1.0.15-singularity.pack.gz # Singularity
  7f60bb29dfb498c1059a33e10e99ac02  vcflib-static-git-v1.0.15-2-g0d40ff1.tar.gz # Static binaries

## ChangeLog v1.0.14 (20250518)

* CMake: Fixed building static libs and shared libs (the previous version just compiled a static lib)
* Updated zig compile for 0.14 (MichaelK)
* Lots of refactoring (MichaelK)
* CMake: Improved linker speeds by moving some libs
* CMake: Improved markdown/pydoc tests, so we don't show the path to the binary
* guix.scm has multiple outputs including static binaries and a test mode
* guix.scm has procedure for building with local htslib and tabixpp only
* Added @RedPenguin100 Michael Kovaliov as a maintainer (welcome!)

## ChangeLog v1.0.13 (20250417)

* Zig support was added to Guix and Conda - and we expect it to be more common soon
* Addind release protocol to README - so I don't forget bumping version etc.
* Fixed vcfwave issues (see https://github.com/vcflib/vcflib/issues/416) and conda build (thanks Michael Kovaliov, aka @RedPenguin100)
* Fixed github actions (Michael)
* Lots of minor refactoring and cleanup, as well as a memory leak (Michael)
* tabix.hpp fixed and set `WFA_INCLUDE_DIRS` and `-DBUILD_SHARED_LIBS=ON` (thanks Michael Cho, aka @cho-m)
* Cmake fixes (thanks Adam Novak)
* Updated simde submodule
* Fix CMakeLists.txt not to try to install wfa-lib when it is a package
* Reverted on Canocalize extraction in Variant class because of downstream use, see https://github.com/vcflib/vcflib/issues/417

## ChangeLog v1.0.12 (20241122)

* Improved parsing of INFO and FORMAT lines by @jeizenga in https://github.com/vcflib/vcflib/pull/374
* Support external tabixpp by @adamnovak in https://github.com/vcflib/vcflib/pull/375
* Fix vcfwave.cpp by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/407
* Use regex to find the '..' between postions and replace it with '-' by @debbyku in https://github.com/vcflib/vcflib/pull/405
* Fix tag-fail long option that was overriden by tag-pass in vcffilter.cpp by @Gullumluvl in https://github.com/vcflib/vcflib/pull/404
* `vcfwave`: fix condition to avoid nullifying valid SNPs and MNPs by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/408
* Upgraded Zig support to 0.13.0 by @pjotrp
* Merged multichoose code into vcflib since no one else uses it
* Added sources for canonicalize too
* Moved Fasta.h (fastahack) dependencies from Variant.h into sources
* Updated multichoose and simde modules
* Improved vcfwave support

### New Contributors
* @jeizenga made their first contribution in https://github.com/vcflib/vcflib/pull/374
* @debbyku made their first contribution in https://github.com/vcflib/vcflib/pull/405
* @Gullumluvl made their first contribution in https://github.com/vcflib/vcflib/pull/404

## ChangeLog v1.0.11-pre ()

+ Stopped vendoring wfa2lib by default - so now the Debian build command is
  `cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DZIG=OFF ..`


## ChangeLog v1.0.10 (20240420)

Vcflib maintenance release

+ Fixed vcfwave bugs - thanks Andrea Guarracino!
+ Fixed c++17 compatability - thanks Alex Petty!
+ Fixed tabixpp script - thanks Blaise Li!
+ Improved for vcfwave the wfa2lib build, fixed running tests with upstream build update
+ Removed deprecated binary_function
+ Removed unused files from repo
+ Updated README

## ChangeLog v1.0.9 (20230211)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.8 (20230210)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.7 (20230207)

Vcflib maintenance release - mostly for including in Debian

+ Fixed regression discovered by garguantua_kerr and atille of Debian (thanks!)
+ Added note on bio-vcf in vcffilter doc
+ notes on vcfcreatemulti and backtracking
+ CMake: honour include(GNUInstallDirs) paths (I forgot)

## ChangeLog v1.0.6 (20230129)

Vcflib maintenance release - mostly for including in Debian

+ Fixed zig complaining about leaking memory
+ Added CMake Debian support with -DWFA_GITMODULE=OFF
+ Introduced CMake include(GNUInstallDirs)
+ Successfully built wfa2 using embedded CMakeLists.txt
+ Cleaned up CMakeLists.txt removing comments etc.
+ Reintroduced vcfcreatemulti in legacy mode when ZIG=OFF (for Debian)

## ChangeLog v1.0.5 (20230116)

Vcflib's first *Humpty Dumpty* release: [vcfcreatemulti](./doc/vcfcreatemulti.md) is the natural companion to [vcfwave](./doc/vcfwave.md).

Often variant callers are not perfect.
**vcfwave** with its companion tool **vcfcreatemulti** can take an existing VCF file that contains multiple complex overlapping and even nested alleles and, unlike Humpty Dumpty, take them apart and put them together again.
Thereby, hopefully, creating sane VCF output that is useful for analysis and getting rid of false positives.

We created these tools by including the state-of-the-art [biWFA](https://github.com/smarco/WFA2-lib) wavefront aligner.

src/permuteGPAT++.cpp
/*
    vcflib C++ library for parsing and manipulating VCF files

    Copyright © 2010-2020 Erik Garrison
    Copyright © 2015      Zev N. Kronenberg
    Copyright © 2020      Pjotr Prins

    This software is published under the MIT License. See the LICENSE file.
*/

/*

This program was created at:  Fri Apr 17 14:59:53 2015
This program was created by:  Zev N. Kronenberg


Contact: zev.kronenber@gmail.com

Organization: Unviersity of Utah
    School of Medicine
    Salt Lake City, Utah


The MIT License (MIT)

Copyright (c) <2015> <Zev N. Kronenberg>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


*/
#include <fstream>
#include <vector>
#include <string>
#include <iostream>
#include <cmath>
#include <ctime>
#include <cstdio>
#include <unistd.h>

#include "split.h"

struct options{
  std::string file;
  int npermutation;
  int nsuc;
}globalOpts;

static const char *optString = "f:n:s:";

using namespace std;


//-------------------------------   OPTIONS   --------------------------------
int parseOpts(int argc, char** argv)
    {
    int opt = 0;
    globalOpts.file = "NA";

    globalOpts.nsuc         = 1;
    globalOpts.npermutation = 1000;

    opt = getopt(argc, argv, optString);
    while(opt != -1){
      switch(opt){
      case 'f':
	{
	  globalOpts.file =  optarg;
	  break;
	}
      case 'n':
	{
	  globalOpts.npermutation = atoi(((string)optarg).c_str());
	  cerr << "INFO: permuteGPAT++ will do N permutations: " << globalOpts.npermutation << endl;
	  break;
	}
      case 's':
	{
	  globalOpts.nsuc = atoi(((string)optarg).c_str());
	  cerr << "INFO: permuteGPAT++ will stop permutations after N successes: " << globalOpts.nsuc << endl;
	  break;
	}
      case '?':
	{
	  break;
	}
      }

      opt = getopt( argc, argv, optString );
    }
    return 1;
    }
//------------------------------- SUBROUTINE --------------------------------
/*
 Function input  : NA

 Function does   : prints help

 Function returns: NA

*/
void printHelp()
{

  cerr << endl << endl;
  cerr << "INFO: help" << endl;
  cerr << "INFO: description:" << endl;
  cerr << "     permuteGPAT++ is a method for adding empirical p-values to a GPAT++ score." << endl << endl ;
  cerr << "     Currently permuteGPAT++ only supports wcFst, but will be extended.    " << endl ;
  cerr << endl;
  cerr << "OUTPUT: permuteGPAT++ will append three additional columns:" << endl;
  cerr << "        1. The number of successes                         " << endl;
  cerr << "        2. The number of trials                            " << endl;
  cerr << "        3. The empirical p-value                           " << endl << endl;

  cerr << "INFO: usage:  permuteGPAT++ -f gpat.txt -n 5 -s 1 "<< endl;
  cerr << endl;
  cerr << "INFO: file:    f   -- argument: the input file     "<< endl;
  cerr << "INFO: number:  n   -- argument: the number of permutations to run for each value [1000]" << endl;
  cerr << "INFO: success: s   -- argument: stop permutations after \'s\' successes [1]"             << endl;

  cerr << endl << "Type: phenotype" << endl << endl;
  cerr << endl;

}

//-------------------------------    MAIN     --------------------------------
/*
 Comments:
*/

int main( int argc, char** argv)
{
  if (argc == 2) {
    string h_flag = argv[1];

    if (argc == 2 && (h_flag == "-h" || h_flag == "--help")) {
      printHelp();
      exit(1);
    }
  }
  int parse = parseOpts(argc, argv);
 if(globalOpts.file.compare("NA") == 0){
   printHelp();
   exit(1);
 }


 vector<double> data;

 ifstream gpat (globalOpts.file.c_str());

 string line;

 if(gpat.is_open()){

   while(getline(gpat, line)){
     vector<string> region = split(line, "\t");
     // will change for other output
     double fst = atof(region[4].c_str());

     if(fst < 0){
       fst = 0;
     }

     data.push_back(fst);
   }
 }
 else{
   cerr << "FATAL: could not open file: " << globalOpts.file << endl;
   exit(1);
 }

 gpat.clear();
 gpat.seekg(0, gpat.beg);

 cerr << "INFO: read values to permute: " << data.size() << endl;

 srand (time(NULL));

 if(gpat.is_open()){

   while(getline(gpat, line)){
     vector<string> region = split(line, "\t");

     double value = atof(region[4].c_str());

     if(value < 0){
       value = 0;
     }


     double suc   = 0;
     double per   = 0;
     int    datas = data.size();
     double pv = (1.0 / globalOpts.npermutation);

     while( suc < globalOpts.nsuc && per < globalOpts.npermutation){
       per += 1.0;

       int r = rand() % datas;

       if(value < data[r]){
	 suc += 1;
       }
     }
     if(suc > 0){
       pv = suc / per;
     }
     cout << line << "\t" << suc << "\t" << per << "\t" << pv << endl;
   }

 }
 else{
   cerr << "FATAL: could not open file: " << globalOpts.file << endl;
   exit(1);
 }


return 0;
}
